First bit of fixing temporary to have one source of truth

abadger opened the pull request:
##### SUMMARY
The change that bcoca said was okay before PTO wasn't okay.  He clarified how things are supposed to work and this is the alternative that should fix that.

<strike>Note that this still doesn't take care of the problem that @jborean93 was seeing. Have to mess with make_tmp_path() (I think that's what we should do) or copy (which calls make_tmp_path and so is the specific place where the error is being shown) before this is ready.</strike>

Confirmed, that the latest iteration of this fixes the problem that @jborean93 noticed.

##### ISSUE TYPE
 - Bugfix Pull Request

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```

abadger committed:
First bit of fixing temporary to have one source of truth

bcoca commented on lib/ansible/plugins/action/__init__.py:
```diff
@@ -87,10 +91,10 @@ def run(self, tmp=None, task_vars=None):
             raise AnsibleActionFail('check mode and async cannot be used on same task.')
 
-        if not tmp and self._early_needs_tmp_path():
+        if self._early_needs_tmp_path():
```
if tmp is supplied we should use that, not ignore it

abadger committed:
Fix pep8

abadger committed:
Remove explicit make_tmp_path() in copy

The copy action plugin sets TRANSFER_FILES=True so it does not need to
set the temporary directory explicitly; the base class's run() method
will do that for us.

abadger committed:
Fix for calling a module's run when a shell has already created a temp path.

bcoca commented on lib/ansible/plugins/action/include_vars.py:
```diff
@@ -80,6 +86,10 @@ def _set_args(self):
         """ Load yml files recursively from a directory.
         """
+        if tmp is not None:
+            display.warning('tmp is no longer a supported parameter of ActionModule.run().  It has no effect')
```
you dont need to add display, just return warning in result

abadger replied:
I can do that but it adds a lot of boilerplate to those modules.   (Because you need to merge the warnings with whatever's already in results).  Some of these modules already use display.XXX as well.  Perhaps I should just use results for the few modules which don't have display already imported?

abadger committed:
Remember to inform the rest of the world when tempdir is removed

nitzmahone reviewed (commented):
a few thoughts:

* changing all the actions not to pass `tmp` to `super` and/or warn (application seems inconsistent, unless I'm missing something?) seems overkill. So long as all the actions actually call `super` early and properly, I'd say just leave it at that. We can't kill it off the ActionBase API without breaking legacy callers outside our codebase, so might as well just rely on the runtime warning rather than adding a bunch of boilerplate to all the actions and trying to pretend it doesn't exist. Though I'd be OK just quietly changing them all to pass `task_vars` as a kwarg...

* if the default behavior is to use the shell-specified tempdir (only when needed), couldn't we also nuke most of the `tmp=self._connection._shell.tempdir` calls to execute_module/execute_remote_stat too? If the tempdir was created ahead of time by the action, it'll get used properly off the shell plugin anyway IIUC.

I think this is probably the right overall approach, but this PR could probably be mostly deletions IMO.

abadger commented:
For (1) The method there is that some of the action plugins (the ones I think are well-behaved) call super().run() near the top of the run() method.  For those, I let the base class's error handling display the warning.  The ones where I used more code, the super().run() was called near the end of the function (usually just prior to return).  For those, it seemed like I shouldn't wait until the end to ```del tmp``` as that would leave it available for people to trip over.  I could do that, though, and rename the data inside to ```old_tmp_value_do_not_use``` or something.  Then pass that to the action base class.

I could also pass the task_vars as a keyword and leave out tmp.  As you say, that would silently ignore people trying to share tmp without telling them that there was now a better standard way to do that.  I'm okay doing it either way just need to know which one is more acceptable.

* For the second, I thought about doing that too.  If that's acceptable, I think I do like it better.  We'd have to tell people that the right way to reuse a tmpdir between two action plugins is to set it on the _shell.tmpdir of the action module that they're chaining to, not by passing it as a parameter.  Want me to switch those too?

abadger committed:
New strategy for how to warn on passing tmp

Now we just warn when calling the parent class run() early.  If the
module does a late call to the parent run() and doesn't make use of the
temporary directory, then we don't check for the possibility that the
user mistakenly is sending tmp in.  If we truly deprecate this (rather
than ignoring it forever) then we might want to switch back to checking
for someone passing a value in as tmp.

abadger committed:
Remove tmp parameter from _execute_module as well

abadger committed:
Port all action plugins to not send tmp explicitly

This is now handled inside of _execute_module via the
_connection._shell.tempdir attribute.

Also update warnings and docs to tell people to set the attribute
instead of using _execute_module's tmp parameter.

abadger committed:
Always set local tempdir variable
